### ISSUE TITLE
Add a function to convert SyncBatchNorm layers back to BatchNorm Layers

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#41081]
`SyncBatchNorm` provides a classmethod to convert `BatchNorm` layers to `SyncBatchNorm` layers:

https://github.com/pytorch/pytorch/blob/881c1adfcd916b6cd5de91bc343eb86aff88cc80/torch/nn/modules/batchnorm.py#L510-L556

It will be helpful to provide an API to do the reverse conversion as well. 

cc @ayush29feb

**Comment by vadimkantorov** [comment]
A more radical proposal: unite BatchNorm*d / SyncBatchNorm in one module BatchNorm (with option `sync = False|True`). I think now BatchNorm*d are all same except shape checking.

There may be same proposal to unite Conv*d into Conv, MaxPool*d into MaxPool etc. Dimensionality may be defined by shape of kernel_size.

Fusion code may become simpler, several versions of fused modules (that now separately exist ConvBn*d, LinearBn*d etc) would likely be simplified and unified.

I think this design already exists in ONNX.

**Comment by thinking-tower** [comment]
Hey @mrshenli, I noticed that this has the label `topic: bootcamp`. If this is still open, I'd be happy to participate in this!

**Comment by ayush29feb** [comment]
Any updates on this issue? I am trying to integrate QAT into my training pipeline and am running into issues as the SyncBatchNorm isn't converted to QuantizedBatchNorm.

**Comment by ayush29feb** [comment]
cc @jerryzh168

**Comment by jerryzh168** [comment]
cc @vkuzo have you tried to do qat on a model with batchnorm and later convert it to quantized batchnorm module?

**Comment by glenn-jocher** [comment]
Issue #1947 raised in YOLOv5 on this topic, when running inference with SyncBN trained models. 

Is there a status update on this?

**Comment by kapily** [comment]
For those who are having trouble with this issue, here's a hackaround function `revert_sync_batchnorm` that "undos" what "convert_sync_batchnorm" does. This is useful if you train via DDP but then want to use that model in CPU inference. 

I haven't submitted a pull request b/c the original BatchNorm1d, BatchNorm2d, etc information is lost and is instead replaced with `BatchNormXd`. If you are ok with the loss of the `_check_input_dim` validation, this should still be functionally identical. 

```
import torch

class BatchNormXd(torch.nn.modules.batchnorm._BatchNorm):
    def _check_input_dim(self, input):
        # The only difference between BatchNorm1d, BatchNorm2d, BatchNorm3d, etc
        # is this method that is overwritten by the sub-class
        # This original goal of this method was for tensor sanity checks
        # If you're ok bypassing those sanity checks (eg. if you trust your inference
        # to provide the right dimensional inputs), then you can just use this method
        # for easy conversion from SyncBatchNorm
        # (unfortunately, SyncBatchNorm does not store the original class - if it did
        #  we could return the one that was originally created)
        return

    
def revert_sync_batchnorm(module):
    # the is very similar to the function that it is trying to revert:
    # https://github.com/pytorch/pytorch/blob/c8b3686a3e4ba63dc59e5dcfe5db3430df256833/torch/nn/modules/batchnorm.py#L679
    module_output = module
    if isinstance(module, torch.nn.modules.batchnorm.SyncBatchNorm):
        new_cls = BatchNormXd
        module_output = BatchNormXd(module.num_features,
                                               module.eps, module.momentum,
                                               module.affine,
                                               module.track_running_stats)
        if module.affine:
            with torch.no_grad():
                module_output.weight = module.weight
                module_output.bias = module.bias
        module_output.running_mean = module.running_mean
        module_output.running_var = module.running_var
        module_output.num_batches_tracked = module.num_batches_tracked
        if hasattr(module, "qconfig"):
            module_output.qconfig = module.qconfig
    for name, child in module.named_children():
        module_output.add_module(name, revert_sync_batchnorm(child))
    del module
    return module_output
```

**Comment by pietermarsman** [comment]
I got slightly different results when using a `BatchNormXd` that was created by the `revert_sync_batchnorm` function. 

I ran the code on a system with a gpu and placed an `import pdf; pdf.set_trace()` right before the `for name, child ...` statement. 

```python
(Pdb) p module
SyncBatchNorm(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
(Pdb) p module_output
BatchNormXd(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
(Pdb) !t = torch.zeros(256, 256)
(Pdb) p module.to('cuda')(t.to('cuda'))
tensor([[  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857],
        [  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857],
        [  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857],
        ...,
        [  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857],
        [  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857],
        [  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857]],
       device='cuda:0', grad_fn=<NativeBatchNormBackward0>)
(Pdb) p module_output.to('cpu')(t.to('cpu'))
tensor([[-0.0882, -0.0213,  0.1118,  ...,  0.0733,  0.0757, -0.0281],
        [-0.0882, -0.0213,  0.1118,  ...,  0.0733,  0.0757, -0.0281],
        [-0.0882, -0.0213,  0.1118,  ...,  0.0733,  0.0757, -0.0281],
        ...,
        [-0.0882, -0.0213,  0.1118,  ...,  0.0733,  0.0757, -0.0281],
        [-0.0882, -0.0213,  0.1118,  ...,  0.0733,  0.0757, -0.0281],
        [-0.0882, -0.0213,  0.1118,  ...,  0.0733,  0.0757, -0.0281]],
       grad_fn=<NativeBatchNormBackward0>)
```

The output is very different. 

There is a difference in the `training` property of the two modules (in my case). One is in training and the other is not. 

```python
(Pdb) p module.training
False
(Pdb) p module_output.training
True
```

When also copying the training attribute the `module_output` the output becomes the same. 

```python
(Pdb) module_output.training = module.training
(Pdb) p module_output.to('cpu')(t.to('cpu'))
tensor([[  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857],
        [  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857],
        [  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857],
        ...,
        [  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857],
        [  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857],
        [  2.1838,   0.1755,   1.9232,  ...,   0.1860, -10.3654,  -1.3857]],
       grad_fn=<NativeBatchNormBackward0>)

```

For completeness, this is @kapily code snippet with the added copying of the training attribute. 

```python
import torch

class BatchNormXd(torch.nn.modules.batchnorm._BatchNorm):
    def _check_input_dim(self, input):
        # The only difference between BatchNorm1d, BatchNorm2d, BatchNorm3d, etc
        # is this method that is overwritten by the sub-class
        # This original goal of this method was for tensor sanity checks
        # If you're ok bypassing those sanity checks (eg. if you trust your inference
        # to provide the right dimensional inputs), then you can just use this method
        # for easy conversion from SyncBatchNorm
        # (unfortunately, SyncBatchNorm does not store the original class - if it did
        #  we could return the one that was originally created)
        return

    
def revert_sync_batchnorm(module):
    # this is very similar to the function that it is trying to revert:
    # https://github.com/pytorch/pytorch/blob/c8b3686a3e4ba63dc59e5dcfe5db3430df256833/torch/nn/modules/batchnorm.py#L679
    module_output = module
    if isinstance(module, torch.nn.modules.batchnorm.SyncBatchNorm):
        new_cls = BatchNormXd
        module_output = BatchNormXd(module.num_features,
                                               module.eps, module.momentum,
                                               module.affine,
                                               module.track_running_stats)
        if module.affine:
            with torch.no_grad():
                module_output.weight = module.weight
                module_output.bias = module.bias
        module_output.running_mean = module.running_mean
        module_output.running_var = module.running_var
        module_output.num_batches_tracked = module.num_batches_tracked
        module_output.training = module.training
        if hasattr(module, "qconfig"):
            module_output.qconfig = module.qconfig
    for name, child in module.named_children():
        module_output.add_module(name, revert_sync_batchnorm(child))
    del module
    return module_output
```

**Comment by awaelchli** [comment]
Nice! Such a utility function would also be useful for PyTorch Lightning to convert the layers back when fitting is done, and also to enable torch.jit.script on the trained model. https://github.com/PyTorchLightning/pytorch-lightning/issues/11076
